k8s/jenkins/apptest/tester/tests/basic-suite.yaml

```diff
@@ -1,8 +1,250 @@
+      done
+
+      # test1 -- unathenticated curl on http://
+      if ! curl -s -k "http://${IP}:${PORT}/" | grep -q 'Authentication required'; then
+        if [[ $DEBUG -gt 0 ]]; then
+          echo "Test1: authentication required -- failed"
+          curl -s -k "http://${IP}:${PORT}/" 2>&1
+        fi
+        $((FAILED+=1))
+      fi
+
+      # test2 -- authenticated curl on http:// (wrong credentials)
+      if ! curl -uadmin:Invalid -s -k "http://${IP}:${PORT}/" | grep -q 'Error 401 Invalid password/token'; then
+        if [[ $DEBUG -gt 0 ]]; then
+          echo "Test2: invalid credentials -- failed"
+          curl -uadmin:Invalid -s -k "http://${IP}:${PORT}/" 2>&1
+        fi
+        $((FAILED+=1))
+      fi
+
+      # test3 -- authenticated curl on http:// (correct credentials)
+      if ! curl -u "admin:${PASSWORD}" -s -k "http://${IP}:${PORT}/" | grep -q 'SetupWizard'; then
+        if [[ $DEBUG -gt 0 ]]; then
+          echo "Test3: valid credentials -- failed"
+          curl -uadmin:${PASSWORD} -s -k "http://${IP}:${PORT}/" 2>&1
+        fi
+        $((FAILED+=1))
+      fi
+
+      if [[ "${FAILED}" != 0 ]]; then
+        echo
+        echo "Username: admin"
+        echo "Password: ${PASSWORD}"
+        echo "Address: http://${IP}:${PORT}/"
+        echo
+        exit "${FAILED}"
+      fi
+    expect:
+      exitCode:
+        equals: 0
+- name: Jenkins ingress login
+  bashTest:
+    script: |-
+      ### 
+      ### WARNING
+      ### 
+      ### Test should use HTTPS protocol but verify doesn't create all
+      ### resources that are created during real deployment.
+      ### In this case it's TLS Secret.
+      ### 
+      ### WARNING
+      ### 
+
+      ### 
+      ### WARNING
+      ### 
+      ### This test will fail once support for Ingress with TLS but without existing TLS Secret is removed.
+      ### There will be no failback to HTTP only Ingress.
+      ### 
+      ### WARNING
+      ### 
+
+      INGRESS="${APP_INSTANCE_NAME}-jenkins-ui"
+      IP=''
+      PASSWORD=''
+      FAILED=0
+
+      function wait_for_ingress_external_ip {
+        IP=$(kubectl get ingress "${INGRESS}" \
+          --namespace "${NAMESPACE}" \
+          -ojsonpath="{.status.loadBalancer.ingress[0].ip}")
+        if [[ $? -ne 0 ]]; then
+          IP=""
+        fi
+      }
+
+      function get_jenkins_pod {
+        POD=$(kubectl get pod -oname \
+          --namespace "${NAMESPACE}" | \
+          sed -n "/\\/${APP_INSTANCE_NAME}-jenkins-deployment/s.pods\\?/..p")
+        if [[ $? -ne 0 ]]; then
+          POD=""
+        fi
+      }
+
+      function get_jenkins_admin_password {
+        PASSWORD=$(kubectl exec "${POD}" \
+          --namespace "${NAMESPACE}" \
+          cat /var/jenkins_home/secrets/initialAdminPassword)
+        if [[ $? -ne 0 ]]; then
+          PASSWORD=""
+        fi
+      }
+
+      function wait_for_ingress_healthy_state {
+        kubectl describe ingress "${INGRESS}" \
```

Comment:
It think the right way should be to use `--output` param here.

```shell
kubectl get ingress "${APP_INSTANCE_NAME}-jenkins-ui" \
  --namespace "${NAMESPACE}" \
  --output jsonpath='{.metadata.annotations.ingress\.kubernetes\.io/backends}'
```

Also, if you add `jq` command, you can get a real status:

```shell
kubectl get ingress "${APP_INSTANCE_NAME}-jenkins-ui" \
  --namespace "${NAMESPACE}" \
  --output jsonpath='{.metadata.annotations.ingress\.kubernetes\.io/backends}'
  | jq -r '.[]'
```

Going forward, you if the `jq` command will be modify a little bit, the exit code `0` will mean success.

```shell
kubectl get ingress "${APP_INSTANCE_NAME}-jenkins-ui" \
  --namespace "${NAMESPACE}" \
  --output jsonpath='{.metadata.annotations.ingress\.kubernetes\.io/backends}' \
  | jq -e '(.[] == "HEALTHY")'
```

Adding loop for this, we will be able to drop the function.

```shell
# Waits until Ingress is healthy.
until kubectl get ingress "${APP_INSTANCE_NAME}-jenkins-ui" \
  --namespace "${NAMESPACE}" \
  --output jsonpath='{.metadata.annotations.ingress\.kubernetes\.io/backends}' \
  | jq -e '(.[] == "HEALTHY")'
do
  sleep 3
done
```

I also suggest to move this logic to the [`tester.sh`](https://github.com/GoogleCloudPlatform/click-to-deploy/blob/master/k8s/jenkins/apptest/tester/tester.sh) file. Generally, we should start testing after deployment is ready. I'll work to improve k8s tools readiness -- so we will be able to remove this logic completely.

Reply:
Here you have the example of WordPress - https://github.com/GoogleCloudPlatform/click-to-deploy/blob/6a3deed4de567bcef9ead52decbbb6670ff6945f/k8s/wordpress/apptest/tester/tester.sh#L20-L27.

I also created a PR to add this functionality to k8s tools - https://github.com/GoogleCloudPlatform/marketplace-k8s-app-tools/pull/315

Reply:
I think what is implemented is simpler.